src/pylibsshext/scp.pyx

```diff
@@ -131,14 +133,17 @@ cdef class SCP:
             if rc == libssh.SSH_ERROR:
                 raise LibsshSCPException("Failed to start read request: %s" % self._get_ssh_error_str())
 
-            # Read the file
-            rc = libssh.ssh_scp_read(scp, read_buffer, size)
-            if rc == libssh.SSH_ERROR:
-                raise LibsshSCPException("Error receiving file data: %s" % self._get_ssh_error_str())
-
-            py_file_bytes = read_buffer[:size]
             with open(local_file, "wb") as f:
-                f.write(py_file_bytes)
+                remaining_bytes_to_read = size
+                while remaining_bytes_to_read > 0:
+                    # Read the file
+                    read_bytes = libssh.ssh_scp_read(scp, read_buffer, min(remaining_bytes_to_read, read_buffer_size))
+                    if read_bytes == libssh.SSH_ERROR:
+                        raise LibsshSCPException("Error receiving file data: %s" % self._get_ssh_error_str())
+
+                    py_file_bytes = read_buffer[:read_bytes]
+                    f.write(py_file_bytes)
+                    remaining_bytes_to_read -= read_bytes
```

Comment:
I believe that this would be more correct:
```suggestion
                    remaining_bytes_to_read -= len(py_file_bytes)
```
The reason is that the slice operator may return fewer bytes, then you think in case the actual iterable is smaller:
```pycon
>>> a_list = [2, 3, 4]
>>> a_list
[2, 3, 4]
>>> a_list[:0]
[]
>>> a_list[:1]
[2]
>>> a_list[:2]
[2, 3]
>>> a_list[:3]
[2, 3, 4]
>>> a_list[:4]
[2, 3, 4]
>>> a_list[:5]
[2, 3, 4]
>>> a_list[:100500]
[2, 3, 4]
```

Reply:
Not sure if this is an issue as if we could not trust the consistency of the returned value `read_bytes` and the amount of read bytes, we are screwed anyway.

Given that we asked to read maximum of `read_buffer_size` and the buffer should have the same size we should not end up in the situation you describe. And if we do, we would create a different file that was on remote anyway so I would rather propose some defensive sanity check where we would fail instead of trying to save it with something like this. Committed a sanity check for the returned size that it is in the bounds we expect.

Reply:
@Jakuje I see you added a check for getting more bytes. But I was talking about the case when it returns *less* bytes, but we'd be removing more bytes than we actually got from the counter.

Imagine we have a file size of 5 and a buffer of 5 but the first invocation of `ssh_scp_read()` does not fill out the entire buffer and only receives 4 bytes. But we still set `read_buffer_size = read_buffer_size - buffer = 5 - 5 = 0` so after the next iteration never happens, and we never attempt getting that last byte.

Reply:
I got that, but I still do not see how this could happen with the current code nor how the proposed fix would fix that. We are subtracting the `read_bytes` from the `remaining_bytes_to_read` and that should always be equal to the `len(py_file_bytes)` unless the `read_bytes` would be larger than the `read_buffer_size`. Did you mistake the `read_buffer_size` and `read_bytes` variables?

Reply:
Yeah, I got confused. I suppose `len(read_buffer[:read_bytes]) == read_bytes` is always true. Marking this thread as resolved.